### PR TITLE
Fix for input number

### PR DESCRIPTION
**Pull request by KarishmaBothara**
Verified on some input number type components.

Pages - 
1. Generic Asset
2. Extrinsic

### CODE REVIEW

File: packages/react-components/src/InputNumber.tsx [hunk 1]
```diff
@@ -240,7 +243,9 @@ export default function InputNumber (props: Props): React.ReactElement<Props> {
 
     if (event.key.length === 1 && !isPreKeyDown) {
       const { selectionStart: i, selectionEnd: j, value } = event.target as HTMLInputElement;
-      const newValue = `${value.substring(0, i || 0)}${event.key}${value.substring(j || 0)}`;
+      let newValue = `${value.substring(0, i || 0)}${event.key}${value.substring(j || 0)}`;
+      // The following will just remove the 0's on the left and send it to check the regex for decimal and allow user to enter number without having to enter backspace
+      newValue = new BN(newValue).toString();
```
Comment (benxgao) [inline packages/react-components/src/InputNumber.tsx]: the second value gonna override the first one, which equals to `const newValue = new BN(newValue).toString();`?

File: packages/react-components/src/InputNumber.tsx [hunk 2]
```diff
@@ -240,7 +243,9 @@ export default function InputNumber (props: Props): React.ReactElement<Props> {
 
     if (event.key.length === 1 && !isPreKeyDown) {
       const { selectionStart: i, selectionEnd: j, value } = event.target as HTMLInputElement;
-      const newValue = `${value.substring(0, i || 0)}${event.key}${value.substring(j || 0)}`;
+      let newValue = `${value.substring(0, i || 0)}${event.key}${value.substring(j || 0)}`;
```
Comment (benxgao) [inline packages/react-components/src/InputNumber.tsx]: As we agreed on commenting practice  https://github.com/cennznet/apps/pull/3/files#diff-0512f381d37e2f1abc301259b7f7abb9R19 , its better to add `// Commented for easy tracking while rebasing'` and comment the original code? Otherwise its hardly for us to know which piece of code we should keep when polkadot updates their code~

Comment (benxgao) [reply packages/react-components/src/InputNumber.tsx]: Would you mind adding easy tracking comment for this PR as well - https://github.com/cennznet/apps/pull/8/files ? 
If we agreed this approach is too waste of time, we could have a discussion and remove this rule from our PR practice, otherwise, it seems more easier for 'polkadot/apps' long-term maintaining, what do you think?

**Comment by KarishmaBothara** [comment]
Work in progress on this PR.. I will re-request review once done

File: packages/react-components/src/InputNumber.tsx [hunk 3]
```diff
@@ -168,12 +168,18 @@ function getValuesFromString (value: string, si: SiDef | null, props: Props): [s
...
+    let decimalValue;
+    if (value.includes('.')) {
+      decimalValue = value.split('.');
+      value = decimalValue[0];
+    }
+    formatedValue = formatNumber(new BN(value.replace(/,/g, '')));
+    formatedValue = decimalValue ? `${formatedValue}.${decimalValue[1]}` : formatedValue;
   }
+  // The following will just remove the 0 on the left, example 01 -> 1
+  formatedValue = value.length === 2 ? new BN(value).toString() : formatedValue;
```
Comment (benxgao) [inline packages/react-components/src/InputNumber.tsx]: We may need to consider edge situations like `010`, `010.`, `010.0`, `010.1`;
What if using `value.startWith('0')` to do the condition check?

File: packages/react-components/src/InputNumber.tsx [hunk 4]
```diff
@@ -168,12 +168,18 @@ function getValuesFromString (value: string, si: SiDef | null, props: Props): [s
...
-    formatedValue = defaultValue.split('.')[0];
+  if (value && !regex.exec(value)) {
+    let decimalValue;
+    if (value.includes('.')) {
+      decimalValue = value.split('.');
+      value = decimalValue[0];
+    }
+    formatedValue = formatNumber(new BN(value.replace(/,/g, '')));
+    formatedValue = decimalValue ? `${formatedValue}.${decimalValue[1]}` : formatedValue;
   }
```
Comment (benxgao) [inline packages/react-components/src/InputNumber.tsx]: It would be better not to reassign value to local variable if possible, I guess that's one reason why `const` is suggested over `let` , how about something like below:

```js
const valueAsArray = value.split(".");
formatedValue = valueAsArray[1]
  ? `${formatNumber(new BN(valueAsArray[0].replace(/,/g, "")))}.${
      valueAsArray[1]
    }`
  : valueAsArray[0];
```

File: packages/react-components/src/InputNumber.tsx (hunk 3 above)
Comment (benxgao) [reply packages/react-components/src/InputNumber.tsx]: May worth exporting the whole function, and add unit tests to cover possible edge conditions?

File: packages/app-accounts/src/modals/Transfer.tsx [hunk 5]
```diff
@@ -72,7 +72,7 @@ function Transfer ({ className, onClose, recipientId: propRecipientId, senderId:
 
   useEffect((): void => {
     if (senderId && recipientId) {
-      setExtrinsic(api.tx.balances.transfer(recipientId, amount || ZERO));
+      setExtrinsic(api.tx.genericAsset.transfer('16000', recipientId, amount || ZERO));
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply packages/app-accounts/src/modals/Transfer.tsx]: This was just for testing accounts page.. I can revert this change back..

File: packages/react-components/src/InputNumber.tsx (hunk 4 above)
Comment (KarishmaBothara) [reply packages/react-components/src/InputNumber.tsx]: Here, if user enters 113. then valueAsArray[1] would not exist. but if users enters something before '.' decimal (won't happen generally) lets say it becomes 4113. here we still want to format it as 4,113.

File: packages/app-accounts/src/modals/Transfer.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (holygits) [reply packages/app-accounts/src/modals/Transfer.tsx]: can remove in that case

File: packages/react-components/src/InputNumber.tsx (hunk 4 above)
Comment (benxgao) [reply packages/react-components/src/InputNumber.tsx]: Right, the line of `: valueAsArray[0]` should have been like `: formatNumber(new BN(valueAsArray[0].replace(/,/g, "")))`, but not a big deal, depends on you, just some code refactoring suggestions.

Comment (KarishmaBothara) [reply packages/react-components/src/InputNumber.tsx]: In that case for some value like 1334., it will become 1334 and decimal will disappear.

**Comment by benxgao** [comment]
Discussed with @KarishmaBothara , at this stage, we would not add comment to track modifying polkadot original code. 

As unit tests have covered edge cases, LGTM

File: packages/react-components/test/InputNumber.spec.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  });
+
+
+  it('Check number format for 0.10:', done => {
+    const value = '0.10';
+    const formatted = formatInput(value);
+    expect(formatted).toEqual('0.10');
+    done();
+  });
+});
```
Comment (benxgao) [inline packages/react-components/test/InputNumber.spec.ts]: 👍
